### DISCUSSION TITLE
Formula for computing Voltage from GPIO 7 reading

### DISCUSSION THREAD

**Discussion by robertlipe** [ale1800/YULC#4] [Show and tell]
I attached a power supply to the barrel jack. I confirmed that removing the fuse reads zero. I then stepped it up from 5->24VDC while reading GPIO7 to get the voltage.

After fidgeting with the numbers slightly, I decided on a formula of =int(GPIO7/175+0.5) as this isn't exactly a high-res number anyway (you have either 12V or 24v lights; you don't have 19.73V lights...) and the one number that's off one is clearly at the saturation point of the ADC (resistor ladder?) anyway.

Below are my measurements as both CSV and the spreadsheet.

Measured,Power Supply,Calculated,Calc Error,Measured diff from previous,
818,5,5,0,,
985,6,6,0,167,
1151,7,7,0,166,
1324,8,8,0,173,
1497,9,9,0,173,
1686,10,10,0,189,
1852,11,11,0,166,
2026,12,12,0,174,
2196,13,13,0,170,
2374,14,14,0,178,
2541,15,15,0,167,
2717,16,16,0,176,
2900,17,17,0,183,
3079,18,18,0,179,
3255,19,19,0,176,
3469,20,20,0,214,
3688,21,21,0,219,
3947,22,23,1,259,Saturation point. Numbers just get wonky
4095,23,23,0,148,

https://docs.google.com/spreadsheets/d/1beT2YhSuUnc9Gq5pti2fZ4-qmi0VcziiZLTm5YjhgyE/pubhtml?gid=0&single=true

**Comment by ale1800** [comment]
Thanks for pointing it out. According to [this](https://espressif-docs.readthedocs-hosted.com/projects/espressif-esp-iot-solution/en/latest/others/adc_range.html#:~:text=The%20maximum%20effective%20range%20of,button%20or%20battery%20voltage%20detection) documentation, it can effectively read up to 3.1V and the resistor divider I made should go up to 3.2V. So I will scale down to a 3V reading for a 24V power supply to be on the safe side👍

**Comment by robertlipe** [comment]
You're the EE, so it's up to you! I was actually pretty happy with the numbers. I didn't do any averaging (maybe I should) and instead just dialed through the voltage, recorded the results, then tinkered with a formula that came up with no really distressing results:

  private:
  int readVoltage() {
      int current_voltage = int(analogRead(kFuseSenseGpioPin) / 170.0 + 0.5);
      return current_voltage;
  }
  
  I didn't click the chip into any Super Clever modes; I just read the ADC with no setup beyond setting it for an input.   I figured we were MOSTLY checking for blown fuses and flagrantly wrong jumper/switch settings, so I wasn't too worried about millivolt resolution. I didn't even use the ADC APIs as I figured I wasn't sampling audio via DMA or anything, I was just looking for really coarse pass/fail/you'reabouttoblowupyourstrip resolution that we could read once every five seconds.
  
  It's pretty dumb that they've not incorporated these changes into the base IDF in all this time. I looked at the code and it's just not there. 
  
  Perhaps my approach is too naive and there's real work to do on the software side, too, if the goal is to get millivolt (centivolt? :-) precision or something. I took your "I read the pin and multiplied by X" guidance pretty literally and I just sketched in an X that worked.
  
  Are the values such that if someone puts 30V (I know...) on the barrel jacks that we never get more than the safe 3.3V into the chip? Overloaded values are OK, but smoking the S3 would be unfortunate if it can be avoided. I mean, there's always going to be some opportunity for damage when someone does something stupid, but we should be sure that slight violations like an unloaded 24VDC that's just running a little high doesn't result in catastrophic failure.
  
  Once I get the initial code committed (I'm on the road) we can iterate on the values and the techniques used.

**Comment by ale1800** [comment]
I can change the divider to have 3.3V at 25V but not more. The main issue is that the buck converter can't handle more than 25V at the input.
